docs/source/notes/randomness.rst

```diff
@@ -2,75 +2,121 @@ Reproducibility
-nondetermnistic. Because floating-point addition is not perfectly associative
-for floating-point operands, :attr:`atomicAdd` with floating-point operands can
-introduce different floating-point rounding errors on each evaluation, which
-introduces a source of nondeterministic variance (aka noise) in the result.
-
-PyTorch functions that use :attr:`atomicAdd` in the forward kernels include
-:meth:`torch.Tensor.index_add_`, :meth:`torch.Tensor.scatter_add_`,
-:meth:`torch.bincount`.
-
-A number of operations have backwards kernels that use :attr:`atomicAdd`,
-including :meth:`torch.nn.functional.embedding_bag`,
-:meth:`torch.nn.functional.ctc_loss`, :meth:`torch.nn.functional.interpolate`,
-and many forms of pooling, padding, and sampling.
-
-There is currently no simple way of avoiding nondeterminism in these functions.
-
-Additionally, the backward path for :meth:`repeat_interleave` operates
-nondeterministically on the CUDA backend because :meth:`repeat_interleave`
-is implemented using :meth:`index_select`, the backward path for
-which is implemented using :meth:`index_add_`, which is known to operate
-nondeterministically (in the forward direction) on the CUDA backend (see above).
-
-cuDNN
-.....
-When running on the cuDNN backend, two further options must be set::
-
-    torch.backends.cudnn.deterministic = True
-    torch.backends.cudnn.benchmark = False
+    import numpy as np
+    np.random.seed(0)
 
-On some versions of cuDNN and CUDA, RNN and LSTM may have non-deterministic behavior.
+However, some applications and libraries may use NumPy Random Generator objects,
+not the global RNG
+(`<https://numpy.org/doc/stable/reference/random/generator.html>`_), and those will
+need to be seeded consistently as well.
+
+If you are using any other libraries that use random number generators, refer to
+the documentation for those libraries to see how to set consistent seeds for them.
+
+CUDA convolution benchmarking
+-----------------------------
+The cuDNN library, used by CUDA convolution operations, can be a source of nondeterminism
+across multiple executions of an application. When a cuDNN convolution is called with a
+new set of size parameters, an optional feature can run multiple convolution algorithms,
+benchmarking them to find the fastest one. Then, the fastest algorithm will be used
+consistently during the rest of the process for the corresponding set of size parameters.
+Due to benchmarking noise and different hardware, the benchmark may select different
+algorithms on subsequent runs, even on the same machine.
+
+Disabling the benchmarking feature with :code:`torch.backends.cudnn.benchmark = False`
+causes cuDNN to deterministically select an algorithm, possibly at the cost of reduced
+performance.
+
+However, if you do not need reproducibility across multiple executions of your application,
+then it would be beneficial to enable this benchmarking feature with
```

Comment:
Can we kill this paragraph? The value is True by default, right, so users don't actually need to do anything?

If we don't kill it, instead of saying "beneficial" can we clarify by saying "performance might improve"?

Reply:
Actually, looks like it's False by default. I can change it to "performance might improve".